### PR TITLE
Improve `prefer-lookaround` rule to report when there are leading/trailing assertions

### LINKED ISSUE #478: `prefer-lookaround` should ignore `^` and `$`
**Information:**
- ESLint version: 8.27.0
- `eslint-plugin-regexp` version: 7.31.10

**Description**

```js
var str = 'JavaScript'.replace(/Java(Script)/g, 'Type$1')
var str = 'JavaScript'.replace(/Java(Script$)/g, 'Type$1')
var str = 'JavaScript'.replace(/Java(Script)$/g, 'Type$1')
```

The first two are reported, but the 3rd is not.

I think 

```js
var str = 'JavaScript'.replace(/Java(?=Script)$/g, 'Type')
```

should result the same. Will adding support for this case break things?

### PR DESCRIPTION
**Pull request by ota-meshi**
close  #478

### CODE REVIEW

File: lib/rules/prefer-lookaround.ts [hunk 1]
```diff
@@ -172,6 +185,205 @@ function getSideEffectsWhenReplacingCapturingGroup(
...
+        )
+    } else if (element.type === "Group") {
+        if (element.alternatives.length === 0) return false
+        for (const alt of element.alternatives) {
+            if (alt.elements.length !== 1) return false
+            if (!isLeadingAssertion(alt.elements[0])) return false
+        }
+        return true
+    }
+    return false
```
Comment (RunDevelopment) [inline lib/rules/prefer-lookaround.ts:237]: Why are there these requirements for leading and trailing assertions? As far as I see it, any zero-length element will work. E.g. `/foo(bar)(?:^|$|a{0}\b|(?=x)|(?<=y))/` -> `/foo(?=bar(?:^|$|a{0}\b)|(?=x)|(?<=y))/` is a valid transformation. Of course, the regex doesn't make sense, but we have other rules that will report these problems.

File: lib/rules/prefer-lookaround.ts [hunk 2]
```diff
@@ -172,6 +185,205 @@ function getSideEffectsWhenReplacingCapturingGroup(
...
+    }
+
+    return {
+        elements,
+        start,
+        end,
+    }
+
+    /** Convert leading/trailing assertion to lookaround assertion pattern text. */
+    function assertionToLookaroundAssertionPatternText(
```
Comment (RunDevelopment) [inline lib/rules/prefer-lookaround.ts:371]: This function has to make sure that the lookaheads/lookbehinds it unwraps look into the same direction as their parent. E.g. `foo(bar)(?<=z\w+)` -> `foo(barz\w+)` is wrong, it has to be `foo(bar(?<=z\w+))`.

Also, should this rule even unwrap lookarounds? Transforming `(?=foo(?=bar))` -> `(?=foobar)` seems like something a dedicated rule should do. Maybe [`no-trivially-nested-assertion`](https://ota-meshi.github.io/eslint-plugin-regexp/rules/no-trivially-nested-assertion.html) could do it?

**Review by RunDevelopment — COMMENTED** [review]
Thank you for implementing this!

I find 2 things we should talk about.

File: lib/rules/prefer-lookaround.ts (hunk 2 above)
Comment (RunDevelopment) [reply lib/rules/prefer-lookaround.ts:371]: Oh, I didn't see #481. That also works for me :)

File: lib/rules/prefer-lookaround.ts (hunk 1 above)
Comment (ota-meshi) [reply lib/rules/prefer-lookaround.ts:237]: Thank you for your review! 
I didn't come up with an idea to check for zero-length element.
I will change to use that method.

File: lib/rules/prefer-lookaround.ts (hunk 2 above)
Comment (ota-meshi) [reply lib/rules/prefer-lookaround.ts]: > Also, should this rule even unwrap lookarounds?

In my opinion, we need to unwrap the simplest lookarounds. prefer-lookaround rule generate lookaround assertions with autofix.
I think the new report after autofix should combine the assertions made by the rule and autofix them.

```js
var str = 'JavaScript'.replace(/Java(S)(c)(r)(i)(p)(t)/g, 'Type$1$2$3$4$5$6')
// ↓
var str = 'JavaScript'.replace(/Java(S)(c)(r)(i)(p)(?=t)/g, 'Type$1$2$3$4$5')
// ↓
var str = 'JavaScript'.replace(/Java(S)(c)(r)(i)(?=pt)/g, 'Type$1$2$3$4')
// ↓
var str = 'JavaScript'.replace(/Java(S)(c)(r)(?>=ipt)/g, 'Type$1$2$3')
// ↓
// ...
// ↓
var str = 'JavaScript'.replace(/Java(?>=Script)/g, 'Type')
```

Comment (RunDevelopment) [reply lib/rules/prefer-lookaround.ts]: Alright, then let's unwrap simple lookarounds.

File: lib/rules/prefer-lookaround.ts [hunk 3]
```diff
@@ -172,6 +180,190 @@ function getSideEffectsWhenReplacingCapturingGroup(
     }
 }
 
+/** Checks if the given element is a zero length element that can be placed as a leading/trailing element. */
+function isLeadingTrailingElement(element: Element): boolean {
+    if (element.type === "CapturingGroup") return false
```
[the remark this answers is not in the document]
Comment (ota-meshi) [reply lib/rules/prefer-lookaround.ts:185]: You're right! I will change this PR.

**Review by RunDevelopment — APPROVED** [review]
Thank you @ota-meshi!

**Pull request merged**